### PR TITLE
fix(zalo): add SSRF guard on outbound photo URLs [AI-assisted]

### PR DESCRIPTION
**Pull request by pgondhi987**
## Summary

- **Problem:** `sendPhoto` in the Zalo extension forwarded user-supplied media URLs directly to the Zalo Bot API with zero URL validation or SSRF protection. The Zalo server fetches the supplied URL from its infrastructure, making it a Server-Side Request Forgery vector.
- **Why it matters:** An attacker who can influence an outbound media URL can cause the Zalo Bot API server to fetch arbitrary internal URLs — cloud metadata endpoints (AWS/GCP/Alibaba Cloud IMDS), localhost services, or internal network hosts.
- **What changed:** `sendPhoto` (`extensions/zalo/src/api.ts`) now (1) rejects non-parseable URLs, (2) enforces an http/https scheme allowlist, (3) passes the hostname through `resolvePinnedHostnameWithPolicy` with a strict empty `SsrFPolicy` before forwarding. The canonical normalized URL (`parsedPhotoUrl.href`) is sent downstream to prevent IDN/encoding discrepancies between the validated and transmitted forms.
- **What did NOT change:** `sendMessage`, `sendChatAction`, webhook functions, and all other Zalo extension behavior are untouched.

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [ ] Refactor required for the fix
- [ ] Docs
- [x] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [ ] Gateway / orchestration
- [ ] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [x] Integrations
- [ ] API / contracts
- [ ] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- [ ] This PR fixes a bug or regression

## Root Cause (if applicable)

- Root cause: `sendPhoto` was a thin wrapper around `callZaloApi` with no URL validation layer. Other extensions that handle outbound media URLs (Telegram, BlueBubbles, QQBot) each have SSRF policies; the Zalo extension had none.
- Missing detection / guardrail: No unit tests existed for `sendPhoto` URL handling; the function had a single line of implementation.
- Contributing context (if known): Zalo extension was not included when SSRF guards were added to other media-sending extensions.

## Regression Test Plan (if applicable)

- Coverage level that should have caught this:
  - [x] Unit test
  - [ ] Seam / integration test
  - [ ] End-to-end test
  - [ ] Existing coverage already sufficient
- Target test or file: `extensions/zalo/src/api.test.ts`
- Scenario the test should lock in: SSRF guard is invoked before any outbound Zalo API call; blocked hostnames throw before reaching the fetcher; non-HTTP schemes and malformed URLs are rejected early.
- Why this is the smallest reliable guardrail: Tests mock `openclaw/plugin-sdk/ssrf-runtime` at the seam and assert call order — guard must fire before fetcher.
- Existing test that already covers this (if any): None (new coverage).
- If no new test is added, why not: N/A — 4 new tests added.

## User-visible / Behavior Changes

Outbound Zalo photo messages where `photo` is a private/internal URL or a non-HTTP scheme will now throw an error instead of silently forwarding the URL to the Zalo API. Legitimate public HTTPS photo URLs are unaffected.

## Diagram (if applicable)

```text
Before:
sendPhoto(token, {photo: attacker_url}) -> callZaloApi -> {photo: attacker_url} forwarded to Zalo

After:
sendPhoto(token, {photo: url})
  -> parse + scheme check (throws on non-http/https or malformed)
  -> resolvePinnedHostnameWithPolicy (throws on private/blocked IPs)
  -> callZaloApi -> {photo: parsedUrl.href} forwarded to Zalo
```

## Security Impact (required)

- New permissions/capabilities? No
- Secrets/tokens handling changed? No
- New/changed network calls? Yes — `resolvePinnedHostnameWithPolicy` performs a DNS lookup to validate the hostname before the Zalo API call is made.
- Command/tool execution surface changed? No
- Data access scope changed? No
- Risk: DNS lookup adds a small latency to `sendPhoto`. Mitigation: this matches the established pattern used by Telegram, BlueBubbles, and QQBot extensions; latency impact is negligible.

### Environment

- OS: Linux (CI)
- Runtime/container: Node 22+
- Integration/channel: Zalo extension (`extensions/zalo/`)

### Steps

1. Call `sendPhoto` with a private-network URL (e.g. `http://169.254.169.254/...`)
2. Observe that `resolvePinnedHostnameWithPolicy` throws before the fetcher is called
3. Call `sendPhoto` with a `file://` URL — observe scheme rejection before SSRF check
4. Call `sendPhoto` with a valid public HTTPS URL — observe normal forwarding

### Expected

- Private/internal URLs: throw before fetcher is invoked
- Non-HTTP schemes: throw before SSRF check is invoked
- Valid public URLs: forwarded normally using the canonicalized `href` form

### Actual

- All 4 test scenarios pass as expected

## Evidence

- [x] Failing test/log before + passing after
- [ ] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

4 new unit tests in `extensions/zalo/src/api.test.ts` cover the full validation matrix.

## Human Verification (required)

This fix was generated by OpenAI Codex (AI-assisted) and reviewed by Claude across two review rounds:

- Round 1 identified: (1) raw trimmed string sent instead of `parsedPhotoUrl.href`; (2) missing test for parse-failure path.
- Round 2 confirmed both findings resolved: `parsedPhotoUrl.href` is now used; `"rejects non-URL strings"` test added.

- Verified scenarios: scheme rejection, SSRF block, parse failure, valid URL success path
- Edge cases checked: `.href` normalization for IDN hostnames; guard fires before fetcher in all error paths
- What you did **not** verify: live Zalo Bot API integration; DNS resolution behavior of `resolvePinnedHostnameWithPolicy` in production network environment

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

## Compatibility / Migration

- Backward compatible? Yes — only affects calls where `photo` is a private/internal/non-HTTP URL (which were silently broken before)
- Config/env changes? No
- Migration needed? No

## Risks and Mitigations

- Risk: `resolvePinnedHostnameWithPolicy` DNS lookup failure on transient network errors could cause legitimate photo sends to fail.
  - Mitigation: This is the same behavior as Telegram/QQBot/BlueBubbles extensions; callers should retry on transient errors as they would for any network operation.

### CODE REVIEW

**Comment by pgondhi987** [comment]
@codex review

File: extensions/zalo/src/send.ts [hunk 1]
```diff
@@ -63,6 +64,30 @@ function resolveSendContext(options: ZaloSendOptions): {
   return { token, fetcher: resolveZaloProxyFetch(proxy) };
 }
 
+async function resolveZaloPhotoParam(photoUrl: string, options: ZaloSendOptions): Promise<string> {
+  const trimmedPhotoUrl = photoUrl.trim();
+  if (!options.cfg) {
+    return trimmedPhotoUrl;
+  }
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/send.ts]: Addressed in 8873cb6a8a2be8faac981913d5f50b657e218102.

File: extensions/zalo/src/outbound-media.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  token: string;
+  buffer: Buffer;
+  contentType?: string;
+  expiresAt: number;
+};
+
+const hostedZaloMedia = new Map<string, HostedZaloMedia>();
+
+function trimTrailingSlash(value: string): string {
+  return value.endsWith("/") ? value.slice(0, -1) : value;
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/outbound-media.ts]: Addressed in 8873cb6a8a2be8faac981913d5f50b657e218102.

File: extensions/zalo/src/monitor.ts [hunk 3]
```diff
@@ -566,6 +572,10 @@ async function processMessageWithPipeline(params: ZaloMessagePipelineParams): Pr
           runtime,
           core,
           config,
+          webhookUrl: account.config.webhookUrl,
+          webhookPath: account.config.webhookPath,
+          mediaMaxBytes: effectiveMediaMaxMb * 1024 * 1024,
+          canHostMedia: useWebhook === true,
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/monitor.ts]: Addressed in 710ce276dd80f9c14c94d86f43188571a4b91893.

File: extensions/zalo/src/outbound-media.ts [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+  return randomBytes(12).toString("hex");
+}
+
+function createHostedZaloMediaToken(): string {
+  return randomBytes(24).toString("hex");
+}
+
+async function ensureHostedZaloMediaDir(): Promise<void> {
+  await mkdir(ZALO_OUTBOUND_MEDIA_DIR, { recursive: true });
+}
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/outbound-media.ts:42]: Addressed in afaa7df5146535eac218242f2107c2ecb428475b.

File: extensions/zalo/src/outbound-media.ts [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+  webhookUrl: string;
+  webhookPath?: string;
+  maxBytes: number;
+}): Promise<string> {
+  await ensureHostedZaloMediaDir();
+  await cleanupExpiredHostedZaloMedia();
+
+  const media = await loadOutboundMediaFromUrl(params.mediaUrl, {
+    maxBytes: params.maxBytes,
+  });
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/outbound-media.ts:132]: Addressed in afaa7df5146535eac218242f2107c2ecb428475b.

File: extensions/zalo/src/outbound-media.ts [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+  }
+
+  const mediaPath = url.pathname;
+  const prefixIndex = mediaPath.lastIndexOf(ZALO_OUTBOUND_MEDIA_PREFIX);
+  if (prefixIndex < 0) {
+    return false;
+  }
+
+  const routePath = mediaPath.slice(0, prefixIndex + ZALO_OUTBOUND_MEDIA_PREFIX.length);
+  const id = mediaPath.slice(prefixIndex + ZALO_OUTBOUND_MEDIA_PREFIX.length);
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/outbound-media.ts:187]: Addressed in c2e6e3158b52d13b5e6bcf7d7573db971cede27b.

File: extensions/zalo/src/monitor.ts [hunk 7]
```diff
@@ -705,6 +757,25 @@ export async function monitorZaloProvider(options: ZaloMonitorOptions): Promise<
         await webhookCleanupPromise;
       };
       runtime.log?.(`[${account.accountId}] Zalo webhook registered path=${path}`);
+      const unregisterHostedMediaRoute = registerPluginHttpRoute({
+        auth: "plugin",
+        match: "prefix",
+        path: hostedMediaRoutePath,
+        replaceExisting: true,
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/monitor.ts]: Addressed in c2e6e3158b52d13b5e6bcf7d7573db971cede27b.

File: extensions/zalo/src/outbound-media.ts [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+  maxBytes: number;
+  fetcher?: HostedZaloMediaFetch;
+}): Promise<string> {
+  await ensureHostedZaloMediaDir();
+  await cleanupExpiredHostedZaloMedia();
+
+  const media = await loadOutboundMediaFromUrl(params.mediaUrl, {
+    maxBytes: params.maxBytes,
+    fetchImpl: params.fetcher,
+    trustExplicitProxyDns: params.fetcher ? true : undefined,
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/outbound-media.ts]: Addressed in c54eb77224c6d85b2a3bcefcb1f82e3a8840af1b.

File: extensions/zalo/src/monitor.ts [hunk 9]
```diff
@@ -566,6 +642,10 @@ async function processMessageWithPipeline(params: ZaloMessagePipelineParams): Pr
           runtime,
           core,
           config,
+          webhookUrl: account.config.webhookUrl,
+          webhookPath: account.config.webhookPath,
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/monitor.ts]: Addressed in c54eb77224c6d85b2a3bcefcb1f82e3a8840af1b.

File: extensions/zalo/src/monitor.ts [hunk 10]
```diff
@@ -174,6 +244,9 @@ function startPollingLoop(params: ZaloPollingLoopParams) {
     config,
     runtime,
     core,
+    canHostMedia,
+    webhookUrl,
+    webhookPath,
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/monitor.ts:256]: Addressed in ff743846aae7200644ba42cfde14faf4b3bb8322.

File: extensions/zalo/src/monitor.ts [hunk 11]
```diff
@@ -566,6 +650,11 @@ async function processMessageWithPipeline(params: ZaloMessagePipelineParams): Pr
           runtime,
           core,
           config,
+          webhookUrl: params.webhookUrl,
+          webhookPath: params.webhookPath,
+          proxyUrl: account.config.proxy,
+          mediaMaxBytes: params.mediaMaxMb * 1024 * 1024,
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/monitor.ts:666]: Addressed in ff743846aae7200644ba42cfde14faf4b3bb8322.

File: extensions/zalo/src/monitor.ts [hunk 12]
```diff
@@ -614,7 +724,17 @@ async function deliverZaloReply(params: {
       }
     },
     sendMedia: async ({ mediaUrl, caption }) => {
-      await sendPhoto(token, { chat_id: chatId, photo: mediaUrl, caption }, fetcher);
+      if (!canHostMedia || !webhookUrl) {
+        throw new Error("Zalo outbound media requires webhook mode with a configured webhookUrl");
+      }
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/monitor.ts]: Addressed in f7ab0a970b2fc9508f653e260a3efbe284292b81.

File: extensions/zalo/src/send.ts [hunk 13]
```diff
@@ -63,6 +64,32 @@ function resolveSendContext(options: ZaloSendOptions): {
...
+    throw new Error("Zalo photo sends require a configured webhookUrl to host media safely");
+  }
+
+  const mediaMaxBytes = (account.config.mediaMaxMb ?? 5) * 1024 * 1024;
+  const proxyUrl = options.proxy ?? account.config.proxy;
+  return await prepareHostedZaloMediaUrl({
+    mediaUrl: trimmedPhotoUrl,
+    webhookUrl,
+    webhookPath: account.config.webhookPath,
+    maxBytes: mediaMaxBytes,
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/send.ts]: Addressed in 0235021a103d7bfe641b62323134b98965ccf490.

**Comment by pgondhi987** [comment]
@codex review

File: extensions/zalo/src/monitor.ts [hunk 14]
```diff
@@ -67,25 +73,84 @@ type ZaloProcessingContext = {
...
+function registerSharedHostedMediaRoute(params: {
+  path: string;
+  accountId: string;
+  log?: (message: string) => void;
+}): () => void {
+  const existing = hostedMediaRouteRefs.get(params.path);
+  if (existing) {
+    existing.count += 1;
+    return () => {
+      const current = hostedMediaRouteRefs.get(params.path);
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/monitor.ts]: Addressed in d1edb85c97e294842e80c3d285a39f9c8fbb2bc3.

File: extensions/zalo/src/outbound-media.ts [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+    webhookPath: params.webhookPath,
+  });
+  const id = createHostedZaloMediaId();
+  const token = createHostedZaloMediaToken();
+  const publicBaseUrl = new URL(params.webhookUrl).origin;
+
+  await writeFile(resolveHostedZaloMediaBufferPath(id), media.buffer, { mode: 0o600 });
+  await writeFile(
+    resolveHostedZaloMediaMetadataPath(id),
+    JSON.stringify({
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/outbound-media.ts]: Addressed in d1edb85c97e294842e80c3d285a39f9c8fbb2bc3.

**Comment by pgondhi987** [comment]
@codex review

File: extensions/zalo/src/monitor.ts [hunk 16]
```diff
@@ -614,7 +726,19 @@ async function deliverZaloReply(params: {
       }
     },
     sendMedia: async ({ mediaUrl, caption }) => {
-      await sendPhoto(token, { chat_id: chatId, photo: mediaUrl, caption }, fetcher);
+      if (!canHostMedia || !webhookUrl || !webhookPath) {
+        throw new Error(
+          "Zalo outbound media requires a configured webhookUrl to host media safely",
+        );
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/monitor.ts]: Addressed in 9055920a371f5c2ef214ed40f3eb7b035aa20cb9.

File: extensions/zalo/src/monitor.ts [hunk 17]
```diff
@@ -67,25 +73,86 @@ type ZaloProcessingContext = {
...
+        res.setHeader("Content-Type", "text/plain; charset=utf-8");
+        res.end("Not Found");
+      }
+    },
+  });
+
+  const existing = hostedMediaRouteRefs.get(params.path);
+  if (existing) {
+    existing.count += 1;
+    existing.unregister = unregister;
```
[the remark this answers is not in the document]
Comment (pgondhi987) [reply extensions/zalo/src/monitor.ts]: Addressed in 9055920a371f5c2ef214ed40f3eb7b035aa20cb9.
